---
output: html_document
---

# Lab 7: RNA Seq Workflow - Gene-level exploratory analysis and differential expression

### Load packages
```{r echo=FALSE, message=FALSE, warning=FALSE, results="asis"}

library("BiocStyle")
library("knitr")
library("rmarkdown")
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE,
               cache = FALSE, fig.width = 5, fig.height = 5)

```

## Reading in data with tximeta

### Load airway data package
```{r}

library("airway")

```

### Determine where files have been installed
```{r}

dir <- system.file("extdata", package="airway", mustWork=TRUE)

```

### List files in the directory
```{r}

list.files(dir)
list.files(file.path(dir, "quants"))

```

### Load sample_table.csv file containing detailed sample information
```{r}

csvfile <- file.path(dir, "sample_table.csv")
coldata <- read.csv(csvfile, row.names=1, stringsAsFactors=FALSE)
coldata

```

### Select first 2 samples in dataset
```{r}

coldata <- coldata[1:2,]
coldata$names <- coldata$Run
coldata$files <- file.path(dir, "quants", coldata$names, "quant.sf.gz")
file.exists(coldata$files)

```

### Load tximeta package and run its main function
```{r message=TRUE}

library("tximeta")
se <- tximeta(coldata)

```

### Determine dimensions of the se object
```{r}

dim(se)
head(rownames(se))

```

### Summarize transcript-level quantifications
```{r message=TRUE}

gse <- summarizeToGene(se)

```

### Check that dimensions are reduced and row IDs are gene IDs
```{r}

dim(gse)
head(rownames(gse))

```

## Summarized Experiment

### Components of a SummarizedExperiment object
The assay (pink block) contains the matrix of counts, the rowRanges (blue block) contains information about the genomic ranges and the colData (green block) contains information about the samples. The highlighted line in each block represents the first row (note that the first row of colData lines up with the first column of the assay).

```{r echo=FALSE}

par(mar=c(0,0,0,0))
plot(1,1,xlim=c(0,100),ylim=c(0,100),bty="n",
     type="n",xlab="",ylab="",xaxt="n",yaxt="n")
polygon(c(45,90,90,45),c(5,5,70,70),col="pink",border=NA)
polygon(c(45,90,90,45),c(68,68,70,70),col="pink3",border=NA)
text(67.5,40,"assay(s)")
text(67.5,35,'e.g. "counts", ...')
polygon(c(10,40,40,10),c(5,5,70,70),col="skyblue",border=NA)
polygon(c(10,40,40,10),c(68,68,70,70),col="skyblue3",border=NA)
text(25,40,"rowRanges")
polygon(c(45,90,90,45),c(75,75,95,95),col="palegreen",border=NA)
polygon(c(45,47,47,45),c(75,75,95,95),col="palegreen3",border=NA)
text(67.5,85,"colData")

```

```{r}

data(gse)
gse

```

### Examine assay names with the function assayNames()
### Examine counts using the function assay() [Note: counts is the first matrix]

```{r}

assayNames(gse)
head(assay(gse), 3)
colSums(assay(gse))

```

### Examine first and last 5 genes using the function rowRanges()
```{r}

rowRanges(gse)

```

### Examine metadata about the sequences using the function seqinfo()
```{r}

seqinfo(rowRanges(gse))

```

### Examine the dataframe that was provided to the tximeta() function using the function colData()
```{r}

colData(gse)

```


## The DESeqDataSet object, sample information, and the design formula

### Examine the columns of gse using $

```{r}

gse$donor
gse$condition

```

### Rename column names 
```{r}

gse$cell <- gse$donor
gse$dex <- gse$condition

```

### Change name of levels
```{r}

levels(gse$dex)
# when renaming levels, the order must be preserved!
levels(gse$dex) <- c("untrt", "trt")

```

### Change the reference group using the function relevel()
```{r}
# %<>% is a compound assignment pipe operator from magrittr, used for consiceness
library("magrittr")
gse$dex %<>% relevel("untrt")
gse$dex

```

```{r eval = FALSE}

#  gse$dex <- relevel(gse$dex, "untrt")

```

## Starting from *SummarizedExperiment*

### Tell colSums argument how many decimals to keep with the function round()
```{r}

round( colSums(assay(gse)) / 1e6, 1 )

```

### Load the package DESeq2
```{r}

library("DESeq2")

```


### Construct a DESeqDataSet object using the function DESeqDataSet()
```{r}

dds <- DESeqDataSet(gse, design = ~ cell + dex)

```

## Starting from count matrices

### Extract only the count matrix and sample info
```{r}

countdata <- round(assays(gse)[["counts"]])
head(countdata, 3)

```

### Manually check that the columns of the count matrix match the rows of the sample information table
```{r}

coldata <- colData(gse)

```

### Construct the DESeqDataSet object from the matrix of counts and sample information using the function DESeqDataSetFromMatrix()
```{r}

ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData = coldata,
                                 design = ~ cell + dex)

```

## Exploratory analysis and visualization
## Pre-filtering the dataset
### Filter rows where counts are greater than 1 to improve speed of functions
```{r}

nrow(dds)
keep <- rowSums(counts(dds)) > 1
dds <- dds[keep,]
nrow(dds)

```

### Filter for rows that contain at least 3 samples of 10 or higher
```{r}

# at least 3 samples with a count of 10 or higher
keep <- rowSums(counts(dds) >= 10) >= 3

```

## The variance stabilizing transformation and the rlog
### Plot standard deviation of genes against the mean Poisson counts
```{r}

lambda <- 10^seq(from = -1, to = 2, length = 1000)
cts <- matrix(rpois(1000*100, lambda), ncol = 100)
library("vsn")
meanSdPlot(cts, ranks = FALSE)

```

### Plot standard deviation of genes against the mean logarithm-transformed counts 
```{r}

log.cts.one <- log2(cts + 1)
meanSdPlot(log.cts.one, ranks = FALSE)

```

### Transform data with a *variance stabilizing transformation* using the function vst()
```{r}

vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)
colData(vsd)

```

### Transform data with a *regularized-logarithm transformation* using the function rlog()
```{r}

rld <- rlog(dds, blind = FALSE)
head(assay(rld), 3)

```

### Visualize the effects of the transformation
```{r fig.width = 6, fig.height = 2.5}

library("dplyr")
library("ggplot2")

dds <- estimateSizeFactors(dds)

df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
         mutate(transformation = "log2(x + 1)"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"))
  
colnames(df)[1:2] <- c("x", "y")  

lvls <- c("log2(x + 1)", "vst", "rlog")
df$transformation <- factor(df$transformation, levels=lvls)

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation) 

```

## Sample distances
### Calculate the Euclidean distance between samples using the function dist()
```{r}

sampleDists <- dist(t(assay(vsd)))
sampleDists

```

### Upload the packages pheatmap and RColorBrewer
```{r}

library("pheatmap")
library("RColorBrewer")

```

### Plot the Euclidean distances on a heatmap using the function pheatmap()
```{r fig.width = 6.1, fig.height = 4.5}

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$dex, vsd$cell, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)

```

### Calculate sample distances with Poisson distance using function PoissonDistance() in the PoiClaClu package
```{r}

library("PoiClaClu")
poisd <- PoissonDistance(t(counts(dds)))

```

### Plot Poisson sample distances on a heatmap using the function pheatmap()
```{r fig.width = 6.1, fig.height = 4.5}

samplePoisDistMatrix <- as.matrix( poisd$dd )
rownames(samplePoisDistMatrix) <- paste( dds$dex, dds$cell, sep=" - " )
colnames(samplePoisDistMatrix) <- NULL
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = colors)

```

## PCA plot
### Visualize sample-to-sample distances using principal component analysis via the function plotPCA()
```{r fig.width=6, fig.height=4.5}

plotPCA(vsd, intgroup = c("dex", "cell"))

```

### Build PCA plot using the ggplot2 package
```{r}

pcaData <- plotPCA(vsd, intgroup = c( "dex", "cell"), returnData = TRUE)
pcaData
percentVar <- round(100 * attr(pcaData, "percentVar"))

```

### Visualize sample-to-sample distances with color representing treatment and shape representign the cell line
```{r fig.width=6, fig.height=4.5}

ggplot(pcaData, aes(x = PC1, y = PC2, color = dex, shape = cell)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA with VST data")

```

## PCA plot using Generalized PCA
### Perform dimension reduction on data using generalized PCA using the function glmpca() from the package glmpca
```{r}

library("glmpca")
gpca <- glmpca(counts(dds), L=2)
gpca.dat <- gpca$factors
gpca.dat$dex <- dds$dex
gpca.dat$cell <- dds$cell

```

### Visualize generalized PCA plot
```{r fig.width=6, fig.height=4.5}

ggplot(gpca.dat, aes(x = dim1, y = dim2, color = dex, shape = cell)) +
  geom_point(size =3) + coord_fixed() + ggtitle("glmpca - Generalized PCA")

```

## MDS plot
### Use multidimensional scaling (MDS) for the VST distances and then visualize them in a plot
```{r fig.width=6, fig.height=4.5}

mds <- as.data.frame(colData(vsd))  %>%
         cbind(cmdscale(sampleDistMatrix))
ggplot(mds, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with VST data")

```

### Use multidimensional scaling (MDS) for the PoissonDistance distances and then visualize them in a plot
```{r fig.width=6, fig.height=4.5}

mdsPois <- as.data.frame(colData(dds)) %>%
   cbind(cmdscale(samplePoisDistMatrix))
ggplot(mdsPois, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with PoissonDistances")

```

## Differential expression analysis
## Running the differential expression pipeline
### Run the differential expression pipeline on raw counts using the function DESeq()
```{r}

dds <- DESeq(dds)

```

## Building the results table
### Extract the estimated log2 fold changes and p values for the last variable in the design formula using the function results()
```{r}

res <- results(dds)
res

```

### Extract the estimated log2 fold changes and p values for only the two levels of dex treatment using contrast argument
```{r}

res <- results(dds, contrast=c("dex","trt","untrt"))

```

### Show metadata of the previous created DataFrame object using the function mcols()
```{r}

mcols(res, use.names = TRUE)

```

### Summarize the results using the function summary()
```{r}

summary(res)

```

### Lower the false discovery rate (FDR) to 0.05
```{r}

res.05 <- results(dds, alpha = 0.05)
table(res.05$padj < 0.05)

```

### Raise the log2 fold change threshold to show effects of treatment on gene counts more than doubling by specifying lfcThreshold = 1
```{r}

resLFC1 <- results(dds, lfcThreshold=1)
table(resLFC1$padj < 0.1)

```

**Note: you can find the citations of packages by typing citation("pkgName") in R. You should cite packages if, in a publication, you use results derived from them.**

## Other comparisons
### Extract results for the log2 of the fold change of one cell line over another
```{r}

results(dds, contrast = c("cell", "N061011", "N61311"))

```

## Multiple testing
### Determine the number of genes with p value < 0.05 and total number of genes tested
```{r}

sum(res$pvalue < 0.05, na.rm=TRUE)
sum(!is.na(res$pvalue))

```

### Determine number of genes with adjust p value < 0.1 as determined using the Benjamini-Hochberg adjustment for multiple testing
```{r}

sum(res$padj < 0.1, na.rm=TRUE)

```

### Subset the results table to the above genes and sort by log2 fold change estimate to rank significant genes starting with those most down-regulated
```{r}

resSig <- subset(res, padj < 0.1)
head(resSig[ order(resSig$log2FoldChange), ])

```

### Same thing as above EXCEPT ranking genes starting with those that are most upregulated
```{r}

head(resSig[ order(resSig$log2FoldChange, decreasing = TRUE), ])

```

## Plotting results
## Counts plot
### Visualize the counts for a particular gene using the function plotCounts()
```{r}

topGene <- rownames(res)[which.min(res$padj)]
plotCounts(dds, gene = topGene, intgroup=c("dex"))

```

### Visualize counts with a customized plot using ggplot()
```{r fig.width = 4, fig.height = 3}

library("ggbeeswarm")
geneCounts <- plotCounts(dds, gene = topGene, intgroup = c("dex","cell"),
                         returnData = TRUE)
ggplot(geneCounts, aes(x = dex, y = count, color = cell)) +
  scale_y_log10() +  geom_beeswarm(cex = 3)

```

```{r fig.width = 4, fig.height = 3}

ggplot(geneCounts, aes(x = dex, y = count, color = cell, group = cell)) +
  scale_y_log10() + geom_point(size = 3) + geom_line()

```

## MA-plot
### Determine the coefficients of the model using the function resultsNames(), shrink the dex_trt_vs_untrt coefficient using the apeglm method, and visualize the distribution of the estimated coefficient using a MA-plot/mean-difference plot/Bland-Altman plot.
```{r fig.width = 4, fig.height = 3}

library("apeglm")
resultsNames(dds)
res <- lfcShrink(dds, coef="dex_trt_vs_untrt", type="apeglm")
plotMA(res, ylim = c(-5, 5))

```


### Plot estimated coefficient without shrinking using a MA-plot
```{r}

res.noshr <- results(dds, name="dex_trt_vs_untrt")
plotMA(res.noshr, ylim = c(-5, 5))

```

### Label an individual point on the MA-plot
```{r}

plotMA(res, ylim = c(-5,5))
topGene <- rownames(res)[which.min(res$padj)]
with(res[topGene, ], {
  points(baseMean, log2FoldChange, col="dodgerblue", cex=2, lwd=2)
  text(baseMean, log2FoldChange, topGene, pos=2, col="dodgerblue")
})

```

### Plot a histogram of the p-values while including only those genes with mean normalized count larger than 1
```{r}

hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")

```

## Gene clustering
### Select the 20 genes with the highest variance across samples from the VST data
```{r}

library("genefilter")
topVarGenes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 20)

```

### Center each genes' values across samples and plot a heatmap
```{r}

mat  <- assay(vsd)[ topVarGenes, ]
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vsd)[, c("cell","dex")])
pheatmap(mat, annotation_col = anno)

```

## Independent filtering
### Create bins using the function quantile(), bin the genes by base mean using the function cut(), rename the levels of the bins using the middle point, calculate the ratio of p values less 0.05 for each bin, and plot the ratios 
**[Note: this process is known as independent filtering, which improves power with multiple testing]**
```{r fig.width=6}

qs <- c(0, quantile(resLFC1$baseMean[resLFC1$baseMean > 0], 0:6/6))
bins <- cut(resLFC1$baseMean, qs)
levels(bins) <- paste0("~", round(signif((qs[-1] + qs[-length(qs)])/2, 2)))
fractionSig <- tapply(resLFC1$pvalue, bins, function(p)
                          mean(p < .05, na.rm = TRUE))
barplot(fractionSig, xlab = "mean normalized count",
                     ylab = "fraction of small p values")

```

## Independent hypothesis weighting
### Instead of independent filtering, weight hypotheses to optimize power using the ihw function in the IHW package
```{r eval=FALSE}

#  library("IHW")
#  res.ihw <- results(dds, filterFun=ihw)

```

## Annotating and exporting results
### Load the packages AnnotationDbi and org.Hs.eg.db
```{r}

library("AnnotationDbi")
library("org.Hs.eg.db")

```

### Get a list of all available key types using the function columns()
```{r}

columns(org.Hs.eg.db)

```

### Use the function mapIds() to add individual columns to the results table
```{r}

ens.str <- substr(rownames(res), 1, 15)
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

```

### Arrange the results by p value and view the head of the DataFrame
```{r}

resOrdered <- res[order(res$pvalue),]
head(resOrdered)

```

## Exporting Results
### Save the results table as a CSV file using the function write.csv()
```{r eval=FALSE}

#  resOrderedDF <- as.data.frame(resOrdered)[1:100, ]
#  write.csv(resOrderedDF, file = "results.csv")

```

### Create a dynamic ReportingTools report which includes links to external databases using gene identifiers and boxplots summarizing the normalized counts across groups
```{r eval=FALSE}

#  library("ReportingTools")
#  htmlRep <- HTMLReport(shortName="report", title="My report",
#                        reportDirectory="./report")
#  publish(resOrderedDF, htmlRep)
#  url <- finish(htmlRep)
#  browseURL(url)

```

## Plotting fold changes in genomic space
### Instead of a DataFrame, return a GRanges object using the argument format and specifying GRanges
```{r}

resGR <- lfcShrink(dds, coef="dex_trt_vs_untrt", type="apeglm", format="GRanges")
resGR

```

### Add symbol again for labeling the genes on the plot
```{r}

ens.str <- substr(names(resGR), 1, 15)
resGR$symbol <- mapIds(org.Hs.eg.db, ens.str, "SYMBOL", "ENSEMBL")

```

### Load the Gviz package
```{r}

library("Gviz")

```

### Specify a window of 1 million base pairs upstream and downstream from the gene with the smallest p value
```{r}

window <- resGR[topGene] + 1e6
strand(window) <- "*"
resGRsub <- resGR[resGR %over% window]
naOrDup <- is.na(resGRsub$symbol) | duplicated(resGRsub$symbol)
resGRsub$group <- ifelse(naOrDup, names(resGRsub), resGRsub$symbol)

```

### Create a vector specifying the genes in the subset having a low value of padj
```{r}

status <- factor(ifelse(resGRsub$padj < 0.05 & !is.na(resGRsub$padj),
                        "sig", "notsig"))

```

### Plot the results using the Gvis functions: GenomeAxisTrack(), AnnotationTrack(), DataTrack(), plotTracks()
```{r}

options(ucscChromosomeNames = FALSE)
g <- GenomeAxisTrack()
a <- AnnotationTrack(resGRsub, name = "gene ranges", feature = status)
d <- DataTrack(resGRsub, data = "log2FoldChange", baseline = 0,
               type = "h", name = "log2 fold change", strand = "+")
plotTracks(list(g, d, a), groupAnnotation = "group",
           notsig = "grey", sig = "hotpink")

```

## Removing hidden batch effects
## Using SVA with DESeq2
### Load the sva package
```{r}

library("sva")

```

### Obtain a matrix of normalized counts for which the average count across samples is > 1, use a full model matrix with the dex variable and a reduced model matrix with only an intercept term, and specify that 2 surrogate variables are wanted
```{r}

dat  <- counts(dds, normalized = TRUE)
idx  <- rowMeans(dat) > 1
dat  <- dat[idx, ]
mod  <- model.matrix(~ dex, colData(dds))
mod0 <- model.matrix(~   1, colData(dds))
svseq <- svaseq(dat, mod, mod0, n.sv = 2)
svseq$sv

```

### Determine how well the SVA method did at recovering the cell line variables
```{r}

par(mfrow = c(2, 1), mar = c(3,5,3,1))
for (i in 1:2) {
  stripchart(svseq$sv[, i] ~ dds$cell, vertical = TRUE, main = paste0("SV", i))
  abline(h = 0)
}

```

### Add the two surrogate variables as columns to the DESeqDataSet
```{r}

ddssva <- dds
ddssva$SV1 <- svseq$sv[,1]
ddssva$SV2 <- svseq$sv[,2]
design(ddssva) <- ~ SV1 + SV2 + dex

```

## Using RUV with DESeq2
### Load the RUVSeq package
```{r}

library("RUVSeq")

```

### Estimate factors of unwanted variation using the function RUVg()
```{r}

set <- newSeqExpressionSet(counts(dds))
idx  <- rowSums(counts(set) > 5) >= 2
set  <- set[idx, ]
set <- betweenLaneNormalization(set, which="upper")
not.sig <- rownames(res)[which(res$pvalue > .1)]
empirical <- rownames(set)[ rownames(set) %in% not.sig ]
set <- RUVg(set, empirical, k=2)
pData(set)

```

### Plot the factors estimated by RUV
```{r}

par(mfrow = c(2, 1), mar = c(3,5,3,1))
for (i in 1:2) {
  stripchart(pData(set)[, i] ~ dds$cell, vertical = TRUE, main = paste0("W", i))
  abline(h = 0)
}

```

### Add factors to the DESeqDataSet as a means to control for them in future analysis
```{r}

ddsruv <- dds
ddsruv$W1 <- set$W_1
ddsruv$W2 <- set$W_2
design(ddsruv) <- ~ W1 + W2 + dex

```

## Time course experiments
### Upload the fission data and designate a design formula that models the strain difference at time 0, the difference over time, and any strain-specific differences over time
```{r}

library("fission")
data("fission")
ddsTC <- DESeqDataSet(fission, ~ strain + minute + strain:minute)

```

### Perform a likelihood ratio test using the function DESeq() indicating LRT for the test argument
```{r}

ddsTC <- DESeq(ddsTC, test="LRT", reduced = ~ strain + minute)
resTC <- results(ddsTC)
resTC$symbol <- mcols(ddsTC)$symbol
head(resTC[order(resTC$padj),], 4)

```

### For the gene with the smallest adjusted p value, plot the counts for the groups over time using ggplot
```{r fig.width=6, fig.height=4.5}

fiss <- plotCounts(ddsTC, which.min(resTC$padj), 
                   intgroup = c("minute","strain"), returnData = TRUE)
fiss$minute <- as.numeric(as.character(fiss$minute))
ggplot(fiss,
  aes(x = minute, y = count, color = strain, group = strain)) + 
  geom_point() + stat_summary(fun.y=mean, geom="line") +
  scale_y_log10()

```

### Perform a wald test for the log2 fold changes at individual time points using the test argument in the function results()
```{r}

resultsNames(ddsTC)
res30 <- results(ddsTC, name="strainmut.minute30", test="Wald")
res30[which.min(resTC$padj),]

```

### Cluster significant genes by their profiles by extracting a matrix of the shrunken log2 fold changes using the function coef()
```{r}

betas <- coef(ddsTC)
colnames(betas)

```

### Plot the log2 fold changes in a heatmap using the function pheatmap()
```{r}

topGenes <- head(order(resTC$padj),20)
mat <- betas[topGenes, -c(1,2)]
thr <- 3 
mat[mat < -thr] <- -thr
mat[mat > thr] <- thr
pheatmap(mat, breaks=seq(from=-thr, to=thr, length=101),
         cluster_col=FALSE)

```

## Session information
### Report the version numbers of R and all the packages used in this session using the function sessionInfo()
**[Note: including this information helps to improve reproducibility and can help troubleshoot issues in the case the R script no longer works]**

```{r}

sessionInfo()

```
